peps/pep-0740.rst

```diff
@@ -205,40 +177,224 @@ changes to it:
+      """
+
+      message_signature: str
+      """
+      The attestation's signature, as `base64(raw-sig)`, where `raw-sig`
+      is the raw bytes of the signing operation.
+      """
+
+  @dataclass
+  class VerificationMaterial:
+      certificate: str
+      """
+      The signing certificate, as `base64(DER(cert))`.
+      """
+
+      transparency_entries: list[TransparencyLogEntry]
+      """
+      Zero or more transparency log entries for this attestation's signature
+      and certificate.
+      """
+
+  @dataclass
+  class TransparencyLogEntry:
+      log_index: int
+      """
+      The global index of the log entry, used when querying the log.
+      """
+
+      log_id: str
+      """
+      An opaque, unique identifier for the log.
+      """
+
+      entry_kind: str
+      """
+      The kind (type) of log entry.
+      """
+
+      entry_version: str
+      """
+      The version of the log entry's submitted format.
+      """
+
+      integrated_time: int
+      """
+      The UNIX timestamp from the log from when the entry was persisted.
+      """
+
+      inclusion_proof: InclusionProof
+      """
+      The actual inclusion proof the the log entry.
+      """
+
+
+  @dataclass
+  class InclusionProof:
+      log_index: int
+      """
+      The index of the entry in the tree it was written to.
+      """
+
+      root_hash: str
+      """
+      The digest stored at the root of the Merkle tree at the time of proof
+      generation.
+      """
+
+      tree_size: int
+      """
+      The size of the Merkle tree at the time of proof generation.
+      """
+
+      hashes: list[str]
+      """
+      A list of hashes required to complete the inclusion proof, sorted
+      in order from leaf to root. The leaf and root hashes are not themselves
+      included in this list; the root is supplied via `root_hash` and the client
+      must calculate the leaf hash.
+      """
+
+      checkpoint: str
+      """
+      The signed tree head's signature, at the time of proof generation.
+      """
+
+Attestation objects are versioned; this PEP specifies version 1. Each version
+is tied to a single cryptographic suite to minimize unnecessary cryptographic
+agility. In version 1, the suite is as follows:
+
+* Certificates are specified as X.509 certificates, and comply with the
+  profile in :rfc:`5280`.
+* The message signature algorithm is ECDSA, with the P-256 curve for public keys
+  and SHA-256 as the cryptographic digest function.
+
+Future PEPs may change this suite (and the overall shape of the attestation
+object) by selecting a new version number.
+
+Attestation payload and signature generation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The *attestation payload* is the actual claim that is cryptographically signed
+over within the attestation object (as the ``message_signature``).
+
+The attestation payload is encoded as an :rfc:`8785` canonicalized JSON object,
+with the following pseudocode layout:
```

Comment:
Flagging: technically the only property we need here is canonicalizability. The PEP currently specifies RFC 8785 since it's a compatible representation of JSON, but I'm not sure if there's a mature implementation of JCS/8785 for Python.

Alternatives here would be DER (since we already have X.509 in the mix) or something bespoke here (e.g. a plain byte string with `:` or `\n` domain separators between fields).

Reply:
Just to dump my research here, here are the JCS implementations I found for Python:

* https://github.com/titusz/jcs (available on PyPI, last updated 2 years ago)
* https://github.com/cyberphone/json-canonicalization/tree/master contains a Python implementation, but not published to PyPI
* https://github.com/AnasMK/json-canonical forked from the above, targets Python 2?
* https://github.com/sveinugu/jsoncanon on PyPI, appears actively maintained, doesn't have FP support (which we probably don't need)

Of these, `jcs` and `jsoncanon` are probably the most immediately usable, but both are relatively small projects.

Reply:
There's also `encode_canonical` from securesystemslib, but AFAICT this doesn't implement JCS -- it implements a different but similarly named canonicalization scheme from OLPC: <https://wiki.laptop.org/go/Canonical_JSON>.

Reply:
Update: I'm working on a minimal implementation of RFC 8785 that should be a maintainable, stable basis for this PEP.

Reply:
I've created https://github.com/trailofbits/rfc8785.py, which should be sufficient for our needs. So I think this is no longer an issue.